pkg/translator/prometheus/normalize_name.go

```diff
@@ -192,6 +194,136 @@ func normalizeName(metric pmetric.Metric, namespace string) string {
 	return normalizedName
 }
 
+func normalizeSingleUnit(unit string, unitMapper func(unit string) string) string {
+	unit = strings.TrimSpace(unit)
+	if unit == "" || strings.ContainsAny(unit, "{}") {
+		return ""
+	}
+
+	return CleanUpString(unitMapper(unit))
+}
+
+func otlpUnitToProm(otlpUnit string) []string {
+	// Split unit at the '/' if any
+	unitTokens := strings.SplitN(otlpUnit, "/", 2)
+	if len(unitTokens) < 1 {
+		return nil
+	}
+
+	// Main unit
+	mainUnit := normalizeSingleUnit(unitTokens[0], unitMapGetOrDefault)
+	if len(unitTokens) == 1 {
+		if mainUnit == "" {
+			return nil
+		}
+
+		return []string{mainUnit}
+	}
+
+	secondaryUnit := normalizeSingleUnit(unitTokens[1], perUnitMapGetOrDefault)
+	if secondaryUnit != "" && mainUnit != "" {
+		return []string{mainUnit, "per", secondaryUnit}
+	}
+
+	// At this point, either mainUnit or secondaryUnit is empty, or both.
+	// Whichever is not empty gets returned, if none, return empty.
+	if mainUnit != "" {
+		return []string{mainUnit}
+	}
+
+	if secondaryUnit != "" {
+		return []string{"per", secondaryUnit}
+	}
+
+	return nil
+}
+
+type Normalizer struct {
+	registry *featuregate.Registry
+}
+
+func NewNormalizer(registry *featuregate.Registry) *Normalizer {
+	return &Normalizer{
+		registry: registry,
+	}
+}
+
+// TrimPromSuffixes trims type and unit prometheus suffixes from a metric name.
+// Following the [OpenTelemetry specs] for converting Prometheus Metric points to OTLP.
+//
+// [OpenTelemetry specs]: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/data-model.md#metric-metadata
+func (n *Normalizer) TrimPromSuffixes(promName string, metricType pmetric.MetricType, unit string) string {
+	if !n.registry.IsEnabled(normalizeNameGateID) {
+		return promName
+	}
+
+	nameTokens := strings.Split(promName, "_")
+	if len(nameTokens) == 1 {
+		return promName
+	}
+
+	nameTokens = removeTypeSuffixes(nameTokens, metricType)
+	nameTokens = removeUnitSuffixes(nameTokens, unit)
+
+	return strings.Join(nameTokens, "_")
+}
+
+func removeTypeSuffixes(tokens []string, metricType pmetric.MetricType) []string {
+	switch metricType {
+	case pmetric.MetricTypeSum:
+		return removeSuffix(tokens, "total")
+	default:
+		return tokens
+	}
+}
+
+func removeUnitSuffixes(nameTokens []string, unit string) []string {
+	l := len(nameTokens)
+
+	if nameTokens[l-1] == unit {
+		// base case, unit is suffixed as is.
+		return nameTokens[:l-1]
+	}
+
+	if unit == "1" {
```

Comment:
note that the UNIT string attached to prometheus metrics should already be prometheus-style units.  We should only be trimming the string off the end, rather than doing any prom->otel conversion (at least when trimming suffixes).  E.g. `http_duration_seconds_total`, with unit `seconds`, becomes `http_duration`.  `http_duration_seconds_total`, with unit `s` is NOT valid OpenMetrics (and only OpenMetrics has unit comments).  We don't need to try and convert `s` to seconds before trimming.

Reply:
What are the expectations for someone using the Otel libraries exposing their metrics with:
1. Language Metric SDK Prometheus exporter -> Collector's Prometheus Receiver -> Collector's OTLP exporter
2. Language Metric SDK OTLP exporter

Should the outcoming OTLP points for these data flow have the same metric name and units or is it not expected to be equivalent?

Reply:
They should have the same metric name and units.

Whenever a unit is transmitted using OTLP, it should have the OTLP representation of the unit (By, s, etc), and should not include unit suffixes.

Whenever a unit is transmitted using prom/OM, it should have the Prom representation of the unit (bytes, seconds, etc) and the unit suffix added. Language SDK prometheus exporters must either convert to prom-style units, or not include unit comments or add suffixes.  The prometheus receiver should blindly trim the incoming unit string from the name suffix, and then convert the unit string to OTLP's unit style.